### PR TITLE
Set valid classloader for RollbackToSnapshotProcedure

### PR DESCRIPTION
**Pull request by homar**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?
a fix
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
connector
> How would you describe this change to a non-technical end user or system administrator?
fixes a bug that didn't allow to run rollback_to_snapshot procedure
## Related issues, pull requests, and links

Fixes: https://github.com/trinodb/trino/issues/12887

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by findinpath** [comment]
Let's mention also #12687  as an issue to which this PR relates to.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        assertThat(onTrino().executeQuery(format("SELECT * FROM %s", tableName)))
+                .containsOnly(row(1));
+    }
+
+    private long getSecondTableSnapshot(String tableName)
+    {
+        return (Long) onTrino().executeQuery(
+                format("SELECT snapshot_id FROM iceberg.default.\"%s$snapshots\" WHERE parent_id IS NOT NULL ORDER BY committed_at LIMIT 1", tableName))
+                .row(0)
+                .get(0);
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:51]: use `getOnlyElement` instead of `row(0)` and `get(0)`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        long snapshotId = getSecondTableSnapshot(tableName);
+        onTrino().executeQuery(format("call system.rollback_to_snapshot('default', '%s', %d)", tableName, snapshotId));
+        assertThat(onTrino().executeQuery(format("SELECT * FROM %s", tableName)))
+                .containsOnly(row(1));
+    }
+
+    private long getSecondTableSnapshot(String tableName)
+    {
+        return (Long) onTrino().executeQuery(
+                format("SELECT snapshot_id FROM iceberg.default.\"%s$snapshots\" WHERE parent_id IS NOT NULL ORDER BY committed_at LIMIT 1", tableName))
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:49]: - Why `parent_id IS NOT NULL`?
- is `ORDER BY committed_at` sufficient for deterministic order? can snapshot be effectively at the same time (within 1ms)? or shall we have a sleep between inserts?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        onTrino().executeQuery(format("CREATE TABLE %s (a INTEGER)", tableName));
+        onTrino().executeQuery(format("INSERT INTO %s VALUES 1", tableName));
+        onTrino().executeQuery(format("INSERT INTO %s VALUES 2", tableName));
+        long snapshotId = getSecondTableSnapshot(tableName);
+        onTrino().executeQuery(format("call system.rollback_to_snapshot('default', '%s', %d)", tableName, snapshotId));
+        assertThat(onTrino().executeQuery(format("SELECT * FROM %s", tableName)))
+                .containsOnly(row(1));
+    }
+
+    private long getSecondTableSnapshot(String tableName)
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:46]: Second -> SecondNewest

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        extends ProductTest
+{
+    @Test(groups = {ICEBERG, PROFILE_SPECIFIC_TESTS})
+    public void testRollbackToSnapshot()
+    {
+        String tableName = "test_rollback_to_snapshot_" + randomTableSuffix();
+
+        onTrino().executeQuery("USE iceberg.default");
+        onTrino().executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
+        onTrino().executeQuery(format("CREATE TABLE %s (a INTEGER)", tableName));
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:37]: drop table at the end

**Review by findepi — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java [hunk 5]
```diff
@@ -64,8 +67,10 @@ public Procedure get()
 
     public void rollbackToSnapshot(ConnectorSession clientSession, String schema, String table, Long snapshotId)
     {
-        SchemaTableName schemaTableName = new SchemaTableName(schema, table);
-        Table icebergTable = catalogFactory.create(clientSession.getIdentity()).loadTable(clientSession, schemaTableName);
-        icebergTable.rollback().toSnapshotId(snapshotId).commit();
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java:70]: Just wondering, pls do not do any implementation yet: considering that this is a fix for the `rollback_to_snapshot` stored procedure call, shouldn't we then apply this method to other places as well in `trino-iceberg` ?

Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java:70]: Could you please add a code comment related to the reasoning behind this change?
This approach is at the moment of this writing rather cryptical to me.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java (hunk 2 above)
Comment (homar) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:49]: > Why parent_id IS NOT NULL?

I wanted to make sure I will not get the first snapshot(its parent id is null) .

>is ORDER BY committed_at sufficient for deterministic order? can snapshot be effectively at the same time (within 1ms)? or shall we have a sleep between inserts?

Sure I can add that sleeping but I have not ever observed two inserts executed one after another to be so fast that  they create  snapshots in the same second. But better safe than sorry I agree

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java (hunk 5 above)
Comment (homar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java:70]: Probably there may be other places where this is necessary but it is not that easy to spot them. 
Our tests cover a lot of scenarios and they seem to work. 
Also not only trino-iceberg is affected. Actually every place that calls `newEmptyConfiguration` can be a problem

Comment (homar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java:70]: @findepi does a great job in explaining this thing, so maybe you should talk to him ;)

Basically in the ticket mentioned by you (https://github.com/trinodb/trino/pull/12687) the check is added that makes sure that only one classloader ever creates new instances of `org.apache.hadoop.conf.Configuration` class. 
When trino is run normally(so not in tests), at the start, it initiates plugins(connectors) and for this uses `PluginClassLoader`. 
Hive, Iceberg and DeltaLake connectors happen to use that `Configuration` class during initialization so it  is `PluginClassLoader` that creates first instances of it and it is stored [here](https://github.com/trinodb/trino/blob/8cb79f08493c949f75b48fc102475ede15e331de/lib/trino-hadoop-toolkit/src/main/java/io/trino/hadoop/ConfigurationInstantiator.java#L42).
However when `rollback_to_snapshot` procedure was called its [rollbackToSnapshot](https://github.com/trinodb/trino/blob/8cb79f08493c949f75b48fc102475ede15e331de/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java#L65)
is called from a Server code not from a connector code and thus different class loader is used `AppClassLoader`.
This change forces using correct class loader, as this class as a whole is part of an iceberg connector so it is loaded/initialized by `PluginClassLoader`, so it is enough to just catch the classLoader used during initialization and use it later. 

Does that make sense? @findepi @findinpath

Comment (homar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java:70]: Btw i will add a code comment but little shorter than this

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java:70]: > Does that make sense?

tl;dr ;)

Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java:70]: > However when rollback_to_snapshot procedure was called its rollbackToSnapshot is called from a server code ...

Does this actually mean that all  (hive/iceberg/delta) stored procedure calls are / may be affected by this problem?

The method `methodHandle` used in the `ROLLBACK_TO_SNAPSHOT` could be probably then adjusted/ slightly duplicated so that it does (under the scenes) the trick you are doing within this class and therefore transparently 
address the problem for the other stored procedure calls that may be uncovered with #12687

```
this.classLoader = Thread.currentThread().getContextClassLoader();
```

Which class loader are we expected to deal with here? Is it the  PluginClassLoader / AppClassLoader/ something else?




> Does that make sense?

Yes, I think I have a better context of the fix now.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java (hunk 1 above)
Comment (homar) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:51]: I am not sure about this. 
I'd probably have to do something like https://github.com/trinodb/trino/blob/83734c37636bc5ab9839535791773244ac0ee0da/testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCompression.java#L104
and not only it is much longer but also intellij gives me warning saying `Unboxing of 'getOnlyElement(getOnlyElement(onTrino().executeQuery( format("SELECT snapshot_id FRO...' may produce 'NullPointerException' 
Argument 'getOnlyElement(onTrino().executeQuery( format("SELECT snapshot_id FROM iceberg.defau...' might be null `

I will do that if you insist but I have doubts

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java (hunk 3 above)
Comment (homar) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergProcedureCalls.java:46]: btw I want second oldest not second newest ;)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java (hunk 5 above)
Comment (homar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RollbackToSnapshotProcedure.java:70]: > Does this actually mean that all (hive/iceberg/delta) stored procedure calls are / may be affected by this problem?

as long as they try to create new instance of Configuration class they may be affected

> The method methodHandle used in the ROLLBACK_TO_SNAPSHOT could be probably then adjusted/ slightly duplicated so that it does (under the scenes) the trick you are doing within this class and therefore transparently
address the problem for the other stored procedure calls that may be uncovered with https://github.com/trinodb/trino/pull/12687

I may be missing something but I don't see how this method could be used. It is static so it is not possible to force it to store the reference to the required class loader.  In theory we could reuse the class loader that loaded `MethodHandleUtil` from which that method comes but we would then need to make sure that each connector initialize that class - I also think that would be equally if not more hacky. 

> Which class loader are we expected to deal with here? Is it the PluginClassLoader / AppClassLoader/ something else?

We need PluginClassLoader here.

**Review by findepi — APPROVED** [review]
